### DISCUSSION TITLE
secp256k1

### DISCUSSION THREAD

**Discussion by Noname400** [exaloop/codon#166] [General]
whether the standard library will work bitcoin. and will it be faster.
how hash functions work ?

**Comment by iuridiniz** [comment]
I think you are talking about using codon to make a [proof of work](https://en.wikipedia.org/wiki/Proof_of_work) miner. In this case codon will be not faster than any other language (CUDA, OpenCL, C++, Go, Rust, etc), the limit is bound to CPU or GPU of the running machine, in other words, the limit is bound to the running **hardware**, not to the language itself. 

Today, the most famous proof of work cryptocurrency, [bitcoin](https://en.wikipedia.org/wiki/Bitcoin) uses [sha-256](https://en.wikipedia.org/wiki/Sha1sum) as [block hashing function](https://en.bitcoin.it/wiki/Block_hashing_algorithm) and it is no more profitable when using CPU/GPU (**hardware**). Due this, people (at date of this message) are using an [ASIC/ASSP](https://en.wikipedia.org/wiki/Application-specific_integrated_circuit) which is a very specialized **hardware** that runs a very narrow set of algorithms. So some ASICs specialized to run sha256 sums (**hardware**) were developed in order to make bitcoin profitable, and every day a new ASIC (**a new hardware**) is developed what obsoletes the old ones (make the old ones less profitable) in an endless cycle.

[Hash functions](https://en.wikipedia.org/wiki/Cryptographic_hash_function) are algorithms, some of they are suitable for parallelism, some not. The fact that Codon can run code in parallel and using GPU's, makes it in pair with any other language that already has native libraries implementing these algorithms, but this do not imply that Codon will be better or faster that any hash libraries already implemented in another language.

**Could Codon implement some hash algorithms in the standard library?** 

Of course, yes. If the goal of this project is to be 100% python compatible, the developers will implement the standard python [`hashlib`](https://docs.python.org/3/library/hashlib.html) someday. But I think this is not the main focus right now.

If you want to make some software using Codon to work with bitcoin and want a sha256 hash function, I suggest you to write a FFI function to [OpenSSL](https://terminalroot.com/how-to-generate-sha256-hash-with-cpp-and-openssl/) or to [libsodium](https://libsodium.gitbook.io/doc/advanced/sha-2_hash_function). The sha256 algorithm itself is not so easy to be implemented purely in codon itself (without FFI), but is not impossible. In this case I suggest you to copy the [pure python implementation of this algorithm provided by PyPy](https://foss.heptapod.net/pypy/pypy/-/blob/branch/default/lib_pypy/_sha256.py) and see if it works in conda without any modification. 

PS: If you clarify your question further, I'll be glad to answer it accordingly.

**Comment by iuridiniz** [comment]
Ok, *now*, I'm seeing the title of your question, you are asking about `Secp256k1` [(this here, right?)](https://en.bitcoin.it/wiki/Secp256k1), so I think you could be talking about wallets and signatures.

I don't know too much about this algorithm, but my words about speed are still valid, the limit is bound to hardware, not to the language. Anything who is already efficient and it is already implement in another language, could not have any real gain in Codon (see FAQ: [Codon is no faster than Python for my application](https://docs.exaloop.io/codon/general/faq#codon-is-no-faster-than-python-for-my-application.)). 

Again, if you want to make some software using Codon, my suggestions are still valid, but I don't know any library or python version of this algorithm to indicate since I don´t know to much about it to **trust** in some random library or code available at google. 

**Trusting** is very important issue here in order to start to use someone else code (regardless of whether it is binary or source), **especially** when it comes to cryptography. Maybe someone put a hidden backdoor in it [(yes, NSA, I'm talking about you)](https://blog.cloudflare.com/how-the-nsa-may-have-put-a-backdoor-in-rsas-cryptography-a-technical-primer/). But if I'm just a lunatic and you are not concern about this, google is your friend [[Pure Python]](https://google.com/search?q=secp256k1+pure+python) / [[Libraries for FFI]](https://google.com/search?q=secp256k1+library).

PS: Again, if you clarify your question further, I'll be glad to answer it accordingly.
